### ISSUE TITLE
✨ LibString escapeHTML

### ISSUE THREAD

**Issue by RogerPodacter** [Vectorized/solady#149]
Now that you're already irritating the "you know, you really shouldn't do this in Solidity" people with `escapeJSON`, why not go all the way? It's also easier because there is no BELL!

Here is a highly-dumb version I wrote long ago that was adapted from the [`escapeQuotes()` method in Uniswap's NFTDescriptor.sol](https://github.com/Uniswap/v3-periphery/blob/main/contracts/libraries/NFTDescriptor.sol#L85). It's not even complete because it doesn't escape quotes, but for my use-case I didn't need to put the string in an HTML attribute which I think is the only reason to escape quotes, but IDK.

```solidity
function escapeHTML(string memory input)
    internal
    pure
    returns (string memory)
{
    bytes memory inputBytes = bytes(input);
    uint extraCharsNeeded = 0;
    
    for (uint i = 0; i < inputBytes.length; i++) {
        bytes1 currentByte = inputBytes[i];
        
        if (currentByte == "&") {
            extraCharsNeeded += 4;
        } else if (currentByte == "<") {
            extraCharsNeeded += 3;
        } else if (currentByte == ">") {
            extraCharsNeeded += 3;
        }
    }
    
    if (extraCharsNeeded > 0) {
        bytes memory escapedBytes = new bytes(
            inputBytes.length + extraCharsNeeded
        );
        
        uint256 index;
        
        for (uint i = 0; i < inputBytes.length; i++) {
            if (inputBytes[i] == "&") {
                escapedBytes[index++] = "&";
                escapedBytes[index++] = "a";
                escapedBytes[index++] = "m";
                escapedBytes[index++] = "p";
                escapedBytes[index++] = ";";
            } else if (inputBytes[i] == "<") {
                escapedBytes[index++] = "&";
                escapedBytes[index++] = "l";
                escapedBytes[index++] = "t";
                escapedBytes[index++] = ";";
            } else if (inputBytes[i] == ">") {
                escapedBytes[index++] = "&";
                escapedBytes[index++] = "g";
                escapedBytes[index++] = "t";
                escapedBytes[index++] = ";";
            } else {
                escapedBytes[index++] = inputBytes[i];
            }
        }
        return string(escapedBytes);
    }
    
    return input;
}
```

Just throwing some ideas out there!

**Comment by Vectorized** [comment]
LMAO. 

Also, shill me a good use case of escapeHTML. 

I feel it may be useful but my brain is not fully working now.

**Comment by RogerPodacter** [comment]
BOOOMMM!

The classic use-case is displaying user-submitted text inside an SVG / HTML element.

This is why I have that code; I used it in my app [Forever Message](https://forevermsg.com) which allows users to create NFTs with their messages. [This one](https://opensea.io/assets/ethereum/0x94f3b9BA57f8ECD2b4aE244e55f2D4F8560fc9c2/29), for example, would have problems if I didn't escape the fake closing SVG tag!

But even when users aren't submitting the text, as is the case in [mega-popular NFT Loot](https://opensea.io/collection/lootproject), it should still be HTML-escaped. One shouldn't rely on the fact that none of their text needs escaping because the rules can be counter-intuitive—e.g., as [I now-famously Tweeted](https://twitter.com/dumbnamenumbers/status/1581295175657132032), an unescaped ampersand that isn't part of a HTML escape sequence (e.g.  "&nbsp;&&nbsp;") is not valid HTML.

In practice, no web browser is going to ding you for an unescaped ampersand. But not all HTML parsers are web browsers and it's not good to rely on their forgiveness! It's good to have valid HTML.

You could store already-escaped HTML, but that's less secure because you might forget to escape something, which is why it's a meme to escape on output instead.

Furthermore, storing text in its original format allows you to use the same data to generate a JSON representation (via `escapeJSON`) as you would use to generate an HTML representation.

In conclusion, `escapeHTML` is a complete slam dunk and, if implemented, will further solidify Solady's epic legacy!

**Comment by Vectorized** [comment]
Ok, Solady express will deliver! Hehe.

**Comment by Vectorized** [comment]
https://github.com/Vectorized/solady/pull/150/files
